providers/amazon/src/airflow/providers/amazon/aws/sensors/s3.py

```diff
@@ -122,8 +122,20 @@ def _check_key(self, key, context: Context):
         """
         if self.wildcard_match:
             prefix = re.split(r"[\[*?]", key, 1)[0]
-            keys = self.hook.get_file_metadata(prefix, bucket_name)
-            key_matches = [k for k in keys if fnmatch.fnmatch(k["Key"], key)]
+
+            key_matches: list[str] = []
+
+            # Is check_fn is None, then we can return True without having to iterate through each value in
+            # yielded by iter_file_metadata. Otherwise, we'll check for a match, and add all matches to the
+            # key_matches list
+            for k in self.hook.iter_file_metadata(prefix, bucket_name):
+                if fnmatch.fnmatch(k["Key"], key):
+                    if self.check_fn is not None:
+                        key_matches.append(k)
+                    # This will only wait for a single match, and will immediately return
+                    else:
+                        return True
```

Comment:
Style "nit" for clarity -- generally try to keep code as un-indented as possible

```suggestion
                    if self.check_fn is None:
                        # This will only wait for a single match, and will immediately return
                        return True
                    key_matches.append(k)
```

Reply:
Thanks - this wasn't my prettiest code :)